File: DTSparkCore/src/main/scala/com/kevin/scala/AccumulatorDemo.scala
package com.kevin.scala

import org.apache.spark.{SparkConf, SparkContext}

/**
  * 累加器在Driver端定义赋初始值和读取，在Executor端累加。
  */
object AccumulatorDemo {

  def main(args: Array[String]): Unit = {

    val file = "DTSparkCore\\src\\main\\resources\\records.txt"
    // 1.创建SparkConf
    val conf = new SparkConf().setAppName("AccumulatorDemo").setMaster("local")
    // 2.创建SparkContext
    val sc = new SparkContext(conf)
    // 3.设置累加器初始值为0
    val accumulator = sc.accumulator(0)
    // 4.textFile读取文件数据,foreach遍历,每遍历一次累加器就加1
    sc.textFile(file).foreach(_ => (accumulator.add(1)))
    // 5.打印累加器的结果
    println(accumulator.value)
    // 6.关闭sc
    sc.stop()
  }

}

File: dt-spark-core/src/main/java/com/kevin/java/AccumulatorDemo.java
package com.kevin.java;

import org.apache.spark.Accumulator;
import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;
import org.apache.spark.api.java.function.VoidFunction;

/**
 * @author kevin
 * @version 1.0
 * @description     累加器在Driver端定义赋初始值和读取，在Executor端累加。
 * @createDate 2019/1/2
 */
public class AccumulatorDemo {

    public static void main(String[] args) {

        SparkConf conf = new SparkConf().setAppName("AccumulatorDemo").setMaster("local");
        JavaSparkContext sc = new JavaSparkContext(conf);

        Accumulator<Integer> accumulator = sc.accumulator(0);

        String file = "DTSparkCore\\src\\main\\resources\\records.txt";
        JavaRDD<String> rdd = sc.textFile(file);

        rdd.foreach(new VoidFunction<String>() {
            @Override
            public void call(String s) throws Exception {
                accumulator.add(1);
            }
        });

        System.out.println(accumulator.value());

        sc.close();
    }
}
